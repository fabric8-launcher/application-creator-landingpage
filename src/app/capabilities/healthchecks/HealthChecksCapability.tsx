import { Grid, GridItem } from '@patternfly/react-core';
import { HeartbeatIcon } from '@patternfly/react-icons';
import * as React from 'react';
import { RequestTitle } from '../../../shared/components/RequestTitle';
import HttpRequest from '../../../shared/components/HttpRequest';
import RequestConsole, { RequestResult } from '../../../shared/components/RequestConsole';
import CapabilityCard from '../../components/CapabilityCard';
import capabilitiesConfig from '../../config/capabilitiesConfig';
import { HealthChecksCapabilityApi, HEALTHCHECKS_LIVENESS_PATH, HEALTHCHECKS_READINESS_PATH } from './HealthChecksCapabilityApi';


interface HealthChecksCapabilityProps {
  apiService: HealthChecksCapabilityApi;
}

interface HealthChecksCapabilityState {
  results: RequestResult[];
}

export default class HealthChecksCapability extends React.Component<HealthChecksCapabilityProps, HealthChecksCapabilityState> {

  constructor(props) {
    super(props);

    this.state = {
      results: [],
    };
  }

  public render() {
    return (
      <CapabilityCard module="healthchecks">
        <CapabilityCard.Title><HeartbeatIcon className="with-text" /> {capabilitiesConfig.healthchecks.name}</CapabilityCard.Title>
        <CapabilityCard.Body>
          <Grid>
            <GridItem span={12}>
              <p>
                Health checks are HTTP endpoints that the codebase exposes for the purpose of monitoring.
                OpenShift will periodically hit these to determine that the container is functioning properly.
                If not, it may kill and/or restart the pod to maintain the targeted number of instances configured to be running.
            </p>
              <p>Your application has been set up with the following Health Checks.</p>
            </GridItem>
            <CapabilityCard.Separator />
            <GridItem span={12} className="http-request-service">
              <RequestTitle>Liveness</RequestTitle>
            </GridItem>
            <HttpRequest method="GET"
              path={HEALTHCHECKS_LIVENESS_PATH}
              curlCommand={`curl -X GET '${this.props.apiService.getLivenessAbsoluteUrl()}'`}
              onExecute={this.execGetLiveness} />
            <GridItem span={12} className="http-request-service">
              <RequestTitle>Readiness</RequestTitle>
            </GridItem>
            <HttpRequest method="GET"
              path={HEALTHCHECKS_READINESS_PATH}
              curlCommand={`curl -X GET '${this.props.apiService.getReadinessAbsoluteUrl()}'`}
              onExecute={this.execGetReadiness} />
            <GridItem span={12}>
              <RequestConsole id="rest-content" results={this.state.results} />
            </GridItem>
          </Grid>
        </CapabilityCard.Body>
      </CapabilityCard>
    );
  }

  private execGetLiveness = async () => {
    try {
      const result = await this.props.apiService.doGetLiveness();
      this.addResult(this.props.apiService.getLivenessAbsoluteUrl(), result);
    } catch (e) {
      this.addResult(this.props.apiService.getLivenessAbsoluteUrl(), {
        time: Date.now(),
        error: 'An error occured while executing the request',
      })
    }
  };

  private execGetReadiness = async () => {
    try {
      const result = await this.props.apiService.doGetReadiness();
      this.addResult(this.props.apiService.getReadinessAbsoluteUrl(), result);
    } catch (e) {
      this.addResult(this.props.apiService.getReadinessAbsoluteUrl(), {
        time: Date.now(),
        error: 'An error occured while executing the request',
      })
    }
  };

  private addResult(url: string, payload: { content?: any, time: number, error?: string }) {
    this.setState({
      results: [...this.state.results, {
        method: 'GET',
        ...payload,
        url,
      }],
    });
  }
}