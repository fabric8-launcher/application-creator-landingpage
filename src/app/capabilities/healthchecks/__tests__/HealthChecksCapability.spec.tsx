import * as React from 'react';
import {shallow} from 'enzyme';
import HealthChecksCapability from '../HealthChecksCapability';
import {MockHealthChecksCapabilityApi} from '../HealthChecksCapabilityApi';
import HttpRequest from '../../../../shared/components/HttpRequest';

const extra = {
  sourceMapping: {
    greetingEndpoint: 'src/main/java/org/your/GreetingApi.java',
  },
}

describe('<HealthChecksCapability />', () => {
  it('check that initial render is correct', () => {
    const component = shallow(<HealthChecksCapability apiService={new MockHealthChecksCapabilityApi()}/>);
    expect(component).toMatchSnapshot();
  });

  it('check that readiness is working', async () => {
    const apiService = new MockHealthChecksCapabilityApi();
    const doGetReadiness = jest.spyOn(apiService, 'doGetReadiness');
    const component = shallow(<HealthChecksCapability apiService={apiService}/>);
    const result = { content: 'OK', time: 1542793377 };
    const promise = Promise.resolve(result);
    doGetReadiness.mockReturnValue(promise);
    component.find(HttpRequest).at(1).prop('onExecute')();
    expect(doGetReadiness).toHaveBeenCalled();
    await promise;
    component.update();
    expect(component.state('results')).toHaveLength(1);
    expect(component.state('results')[0].content).toBe(result.content);
    expect(component).toMatchSnapshot();
  });


});

