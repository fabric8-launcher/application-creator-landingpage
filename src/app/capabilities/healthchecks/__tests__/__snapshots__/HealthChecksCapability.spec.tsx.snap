// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HealthChecksCapability /> check that initial render is correct 1`] = `
<CapabilityCard
  module="healthchecks"
>
  <Component>
    <HeartbeatIcon
      className="with-text"
      color="currentColor"
      size="sm"
      title={null}
    />
     
    Health Checks
  </Component>
  <Component>
    <Grid
      className=""
      gutter={null}
      lg={null}
      md={null}
      sm={null}
      span={null}
      xl={null}
    >
      <GridItem
        className=""
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <p>
          Health checks are HTTP endpoints that the codebase exposes for the purpose of monitoring. OpenShift will periodically hit these to determine that the container is functioning properly. If not, it may kill and/or restart the pod to maintain the targeted number of instances configured to be running.
        </p>
        <p>
          Your application has been set up with the following Health Checks.
        </p>
        <p>
          <ExternalLink
            href="https://docs.openshift.com/container-platform/latest/dev_guide/application_health.html"
          >
            Learn more
          </ExternalLink>
        </p>
      </GridItem>
      <Component />
      <GridItem
        className="http-request-service"
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <RequestTitle>
          Liveness
        </RequestTitle>
      </GridItem>
      <HttpRequest
        curlCommand="curl -X GET 'http://mocked.io/liveness'"
        method="GET"
        onExecute={[Function]}
        path="/health"
      />
      <GridItem
        className="http-request-service"
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <RequestTitle>
          Readiness
        </RequestTitle>
      </GridItem>
      <HttpRequest
        curlCommand="curl -X GET 'http://mocked.io/readiness'"
        method="GET"
        onExecute={[Function]}
        path="/health"
      />
      <GridItem
        className=""
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <RequestConsole
          id="rest-content"
          results={Array []}
        />
      </GridItem>
    </Grid>
  </Component>
</CapabilityCard>
`;

exports[`<HealthChecksCapability /> check that readiness is working 1`] = `
<CapabilityCard
  module="healthchecks"
>
  <Component>
    <HeartbeatIcon
      className="with-text"
      color="currentColor"
      size="sm"
      title={null}
    />
     
    Health Checks
  </Component>
  <Component>
    <Grid
      className=""
      gutter={null}
      lg={null}
      md={null}
      sm={null}
      span={null}
      xl={null}
    >
      <GridItem
        className=""
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <p>
          Health checks are HTTP endpoints that the codebase exposes for the purpose of monitoring. OpenShift will periodically hit these to determine that the container is functioning properly. If not, it may kill and/or restart the pod to maintain the targeted number of instances configured to be running.
        </p>
        <p>
          Your application has been set up with the following Health Checks.
        </p>
        <p>
          <ExternalLink
            href="https://docs.openshift.com/container-platform/latest/dev_guide/application_health.html"
          >
            Learn more
          </ExternalLink>
        </p>
      </GridItem>
      <Component />
      <GridItem
        className="http-request-service"
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <RequestTitle>
          Liveness
        </RequestTitle>
      </GridItem>
      <HttpRequest
        curlCommand="curl -X GET 'http://mocked.io/liveness'"
        method="GET"
        onExecute={[Function]}
        path="/health"
      />
      <GridItem
        className="http-request-service"
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <RequestTitle>
          Readiness
        </RequestTitle>
      </GridItem>
      <HttpRequest
        curlCommand="curl -X GET 'http://mocked.io/readiness'"
        method="GET"
        onExecute={[Function]}
        path="/health"
      />
      <GridItem
        className=""
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <RequestConsole
          id="rest-content"
          results={
            Array [
              Object {
                "content": "OK",
                "method": "GET",
                "time": 1542793377,
                "url": "http://mocked.io/readiness",
              },
            ]
          }
        />
      </GridItem>
    </Grid>
  </Component>
</CapabilityCard>
`;
